sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/ReduceNumShufflePartitions.scala

```diff
@@ -91,14 +93,21 @@ case class ReduceNumShufflePartitions(conf: SQLConf) extends Rule[SparkPlan] {
       val distinctNumPreShufflePartitions =
         validMetrics.map(stats => stats.bytesByPartitionId.length).distinct
       if (validMetrics.nonEmpty && distinctNumPreShufflePartitions.length == 1) {
-        val partitionStartIndices = estimatePartitionStartIndices(validMetrics.toArray)
-        // This transformation adds new nodes, so we must use `transformUp` here.
-        plan.transformUp {
-          // even for shuffle exchange whose input RDD has 0 partition, we should still update its
-          // `partitionStartIndices`, so that all the leaf shuffles in a stage have the same
-          // number of output partitions.
-          case stage: QueryStageExec if ShuffleQueryStageExec.isShuffleQueryStageExec(stage) =>
-            CoalescedShuffleReaderExec(stage, partitionStartIndices)
+        val visitedStage = mutable.HashSet[QueryStageExec]()
+        plan.transformDown {
+          case stage: QueryStageExec if (ShuffleQueryStageExec.isShuffleQueryStageExec(stage)
+            && !visitedStage.contains(stage)) =>
+            val excludedPartitions =
+              ShuffleQueryStageExec.getShuffleStage(stage).excludedPartitions
+            val partitionIndices = estimatePartitionStartAndEndIndices(
```

Comment:
We should only call it once, not once per stage. We should make sure all shuffles in the stage have the same `excludedPartitions`, otherwise we should give up this rule.